lib/spack/spack/solver/concretize.lp

```diff
@@ -1197,45 +1200,36 @@ error(100, "Compiler {1}@{2} requested for {0} cannot be found. Set install_miss
 % Compiler flags
 %-----------------------------------------------------------------------------
 
-% propagate flags when compiler match
-can_inherit_flags(PackageNode, DependencyNode, FlagType)
-  :- same_compiler(PackageNode, DependencyNode),
-     not attr("node_flag_set", DependencyNode, FlagType, _),
-     flag_type(FlagType).
-
 same_compiler(PackageNode, DependencyNode)
   :- depends_on(PackageNode, DependencyNode),
      node_compiler(PackageNode, CompilerID),
      node_compiler(DependencyNode, CompilerID),
      compiler_id(CompilerID).
 
-node_flag_inherited(DependencyNode, FlagType, Flag)
-  :- attr("node_flag_set", PackageNode, FlagType, Flag),
-     can_inherit_flags(PackageNode, DependencyNode, FlagType),
-     attr("node_flag_propagate", PackageNode, FlagType).
+% propagate flags when compiler match
+attr("node_flag_propagate", Package, FlagType, Flag, Source) :-
+  attr("node_flag_possible_prop", Package, FlagType, Flag, Source),
+  not attr("node_flag_set", Package, FlagType, _).
```

Comment:
```suggestion
attr("node_flag_propagate", PackageNode, FlagType, Flag, Source) :-  % source is a node() attribute
  attr("node_flag_possible_prop", PackageNode, FlagType, Flag, Source),
  not attr("node_flag_set", PackageNode, FlagType, _).
```
We have a convention of using `PackageNode` for `node(_, Package)` and `Package` for the package name.